### PR TITLE
Use MIRSG `molecule-test`

### PR DESCRIPTION
**Pull request by p-j-smith**
Reduce duplication in workflows by:
- using our `molecule-test` GitHub Action
- adding a reusable workflow for running the action
- use this reusable workflow for running tests on the `provision` and `firewalld` roles (more workflows can be added in a separate pr)
- run the tests for a given role when either the role itself changes or the workflow changes

Note, this depends on https://github.com/UCL-MIRSG/.github/pull/80 being merged first.
And linting is broken will fix in a different pr

### CODE REVIEW

File: .github/workflows/linting.yml [hunk 1]
```diff
@@ -11,7 +11,7 @@ jobs:
   linting:
     runs-on: ubuntu-latest
     steps:
-      - uses: UCL-MIRSG/.github/actions/linting@v0.36.0
+      - uses: UCL-MIRSG/.github/actions/linting@v0.38.0
```
Comment (paddyroddy) [inline .github/workflows/linting.yml:14]: One day Renovate will come back 😆

File: .github/workflows/molecule.yml [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+      ANSIBLE_FORCE_COLOR: 1
+
+    steps:
+      - name: Run `molecule test`
+        uses: UCL-MIRSG/.github/actions/molecule-test@v0.38.0
+        with:
+          scenario: ${{ matrix.scenario }}
+          checkout_path: ansible_collections/mirsg/infrastructure
+          tests_path: ansible_collections/mirsg/infrastructure/tests
+
```
Comment (paddyroddy) [inline .github/workflows/molecule.yml:30]: missing new line at the end

File: .github/workflows/molecule.yml [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+        default: all
+
+jobs:
+  molecule:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: true
+      matrix:
+        scenario:
+          - centos7
```
Comment (paddyroddy) [inline .github/workflows/molecule.yml:17]: Only `centos7`?

File: .github/workflows/molecule.yml [hunk 4]
```diff
@@ -0,0 +1,30 @@
+name: Test with Molecule
+on:
+  workflow_call:
+    inputs:
+      run_tags:
```
Comment (paddyroddy) [inline .github/workflows/molecule.yml:5]: Generally people use `run-tags` rather than `run_tags`

**Review by paddyroddy — COMMENTED** [review]
Seems a bit odd to use a reusable workflow to call a composite workflow? Perfectly valid, but is it needed?

File: .github/workflows/molecule.yml (hunk 3 above)
Comment (p-j-smith) [reply .github/workflows/molecule.yml:17]: yeah for now, I'll add Rocky 9 in a separate pr

File: .github/workflows/linting.yml (hunk 1 above)
Comment (p-j-smith) [reply .github/workflows/linting.yml:14]: we can only dream

Comment (drmatthews) [reply .github/workflows/linting.yml:14]: Dreams do come true - I just reenabled it...

I put it on the shared calendar to reenable, but I still forgot...

File: .github/workflows/molecule.yml (hunk 4 above)
Comment (p-j-smith) [reply .github/workflows/molecule.yml]: I've renamed it now

**Comment by p-j-smith** [comment]
> Seems a bit odd to use a reusable workflow to call a composite workflow? Perfectly valid, but is it needed?

Yeah it is a little convoluted. The `molecule-test` action that's in our `.github` repo can be used both for testing this collection as well as for testing standalone roles. When testing the roles within this collection there's a bit of boilerplate that's needed (and that we'd use for each role we want to test). So I put this boilerplate into a reusable workflow. We could have a workflow `molecule-test-collection` but that would duplicate a lot of what's in the existing `molecule-test` action. Roundabout way of saying it's not really necessary, just seemed like a sensible way of doing things. Happy to change though!

File: .github/workflows/linting.yml (hunk 1 above)
Comment (paddyroddy) [reply .github/workflows/linting.yml:14]: I was scared to re-enable it haha

File: .github/workflows/molecule.yml (hunk 2 above)
Comment (paddyroddy) [reply .github/workflows/molecule.yml]: Ah, just saw the dreaded 🚫

**Comment by paddyroddy** [comment]
> Roundabout way of saying it's not really necessary, just seemed like a sensible way of doing things. Happy to change though!

Happy with whatever you decide. If it's a duplication concern, I assume one composite action can depend on another (maybe)?

**Comment by p-j-smith** [comment]
> Happy with whatever you decide. If it's a duplication concern, I assume one composite action can depend on another (maybe)?

Yeah we could have another action `molecule-test-collection` that uses the existing `molecule-test` action, and have each workflow use that `molecule-test-collection`. That new action would look like this:

https://github.com/UCL-MIRSG/ansible-collection-infra/blob/0ca6652ae9ac4571dc40d5a05a6cef224cf1d8b3/.github/actions/molecule-test-collection/action.yml#L1-L19

And using it in a workflow:
https://github.com/UCL-MIRSG/ansible-collection-infra/blob/0ca6652ae9ac4571dc40d5a05a6cef224cf1d8b3/.github/workflows/molecule-firewalld-with-action.yml#L1-L26

Compared to the current workflow for the same role:
https://github.com/UCL-MIRSG/ansible-collection-infra/blob/0ca6652ae9ac4571dc40d5a05a6cef224cf1d8b3/.github/workflows/molecule-firewalld.yml#L1-L13

So using a reusable workflow allows us to avoid some boilerplate that can't be avoided when using a composite action

File: .github/workflows/molecule.yml [hunk 5]
```diff
@@ -0,0 +1,29 @@
+name: Test with Molecule
+on:
+  workflow_call:
+    inputs:
+      run-tags:
+        required: false
+        type: string
+        default: all
```
Comment (paddyroddy) [inline .github/workflows/molecule.yml:8]: `required` isn't required if a `default` is set

Comment (p-j-smith) [reply .github/workflows/molecule.yml]: well spotted, I've removed it

**Pull request merged**